.github/workflows/openstack/prune_openstack

```diff
@@ -1,84 +1,87 @@
-        then
-            echo "deleting: $ID,$CREATED_DATE,$NAME"
-            openstack server delete "$ID"
-        fi
-    done < <(cat id_created_name.txt)
-}
+use POSIX ();
+use strict;
+use warnings;
 
-get_keys () {
-    LIST=$(openstack keypair list -f json  | jq -r .[].Name)
-    rm keyid_created_name.txt > /dev/null 2>&1
-    while IFS= read -r KEYPAIR; do
-        openstack keypair show -f json "$KEYPAIR" | jq -r '.id + "," + .created_at + "," + .name' | grep $KEY_NAME >> keyid_created_name.txt
-    done <<< "$LIST"
-}
+# Environment constants
+our $OS_APPLICATION_CREDENTIAL_ID     = $ENV{OS_APPLICATION_CREDENTIAL_ID};
+our $OS_APPLICATION_CREDENTIAL_SECRET = $ENV{OS_APPLICATION_CREDENTIAL_SECRET};
 
-remove_stale_keys () {
-    echo "HAMMERTIME: $HAMMER_TIME"
-    while IFS="," read -r ID CREATED_DATE NAME
-    do
-        if [[ $CREATED_DATE < $HAMMER_TIME ]];
-        then
-            echo "deleting: $ID,$CREATED_DATE,$NAME"
-            openstack keypair delete "$ID"
-        fi
-    done < <(cat keyid_created_name.txt)
-}
+# Static constants
+use constant {
+    OS_AUTH_TYPE            => 'v3applicationcredential',
+    OS_AUTH_URL             => 'https://keystone.hou-01.cloud.prod.cpanel.net:5000/v3',
+    OS_IDENTITY_API_VERSION => '3',
+    OS_REGION_NAME          => 'RegionOne',
+    OS_INTERFACE            => 'public',
+    KEY_NAME                => "deletethis",
+    VM_NAME                 => "elevate.github.cpanel.net",
+};
 
-get_volumes () {
-    rm volume_ids.txt > /dev/null 2>&1
-    openstack volume list | grep packer  | grep available | awk '{print $2}' >> volume_ids.txt
-}
+my $openstack_path = "/usr/bin/openstack" if -x "/usr/bin/openstack";
+die "No Openstack CLI binary installed." unless $openstack_path;
+
+# Define "two hours ago" timestamp
+my $time_offset = 7200;
+my $hammer_time = POSIX::strftime( "%Y-%m-%dT%H:%M:%SZ", gmtime( time - $time_offset ) );
+
+# Run removal functions
+remove_stale_entities( 'keys',      KEY_NAME, 'keypair' );
+remove_stale_entities( 'instances', VM_NAME,  'server' );
+
+# Fetch and format entities
+sub get_entities {
+    my ( $list_command, $show_command ) = @_;
 
-remove_volumes () {
-    while IFS= read -r ID; do
-        openstack volume delete "$ID" || true
-    done < <(cat volume_ids.txt)
+    my $entities = {};
+    my @list     = split( '\n', `$list_command` );
+
+    foreach my $entity_name (@list) {
+        my $created_cmd = qq{ $show_command "$entity_name" | jq -r '.id + "," + .created_at + "," + .name' };
+        my ( $id, $created_on, $name ) = split( ',', _run_cmd( $created_cmd, 0 ) );
+
+        $entities->{$entity_name} = {
+            id         => $id,
+            created_on => $created_on,
+            name       => $name,
+        };
+        chomp( $entities->{$entity_name}{'name'} );
+    }
+    return $entities;
 }
 
-cleanup () {
-    rm -fv id_created_name.txt
-    rm -fv keyid_created_name.txt
+# Remove stale entities
+sub remove_stale_entities {
```

Comment:
FIX: This sub needs to end with an explicit return

Reply:
Fixed